prow/federated_cluster_credentials.md

```diff
@@ -0,0 +1,224 @@
+
+| Term | Definition |
+| ---- | ---------- |
+| _cluster_ | A _cluster_ is a conformant Kubernetes deployment. _clusters_ are specified in a `~/.kube/config` file. |
+| `cluster` | The `cluster` field on a ProwJob refers to the _context_ in which resources are created to execute it. |
+| _context_ | A _context_ is an identifier for a credential, specifying a _cluster_, a `Namespace` on that cluster, and a _user_ whose credentials should be used when interacting. _contexts_ are specified in a `~/.kube/config` file. |
+| _user_    | A _user_ is an identifier for a Kubernetes identity; _users_ are specified in a `~/.kube/config` file. |
+| _service cluster_ | The _service cluster_ is the _context_ in which the `ProwJob` `CustomResourceDefinition` is deployed. |
+| _in-cluster context_ | The _in-cluster context_ is a _context_ for the `ServiceAccount` under which a `Pod` runs, loaded from a customary location in the `Pod`'s filesystem. |
+
+### Flags
+
+All Prow controllers that interact with resources expose the following flags:
+
+| Flag | Use |
+| ---- | --- |
+| `--kubeconfig` | Path to a `~/.kube/config` file to load. Can be passed more than once. |
+| `--use-explicit-credentials` | A feature flag to opt into the behavior in this document. |
+| `--prow-job-context` | The _context_ for the _service cluster_, used to interact with `ProwJob` `CustomResources`. |
+ 
+Note: until January 2020, the behavior in this document will be guarded behind the `--use-explicit-credentials` flag.
+
+## Configuring Credentials
+
+In order to configure federated credentials, Prow uses the `Config` format (often stored in `~/.kube/config`). Multiple `Config` files may be provided by passing `--kubeconfig` more than once. These files will be merged with a simple union on the _clusters_, _contexts_ and _users_ in each. If any names collide, loading will fail.
+
+## Defaulting
+
+When the `--kubeconfig` flag is not passed, Prow will use the _in-cluster context_ for all operations, like:
+
+ - creating and reading `ProwJob` `CustomResources`
+ - running `Pods` to execute `ProwJob`s that use the `kubernetes` agent
+ - executing Tekton pipelines for `ProwJob`s that use the `tekton` agent
+ 
+If a `--kubeconfig` flag is passed and the `--use-explicit-credentials` flag is set, no defaulting of any sort will occur. No _context_ is default and no _context_ is special in this mode. Jobs must opt into using a specific loaded _context_ by setting the `cluster` field. Prow components will continue using the _in-cluster context_ for interacting with `ProwJob` `CustomResources`.
+
+## Specifying an _in-cluster context_ via `--kubeconfig`
+
+Prow will post-process the `Config` loaded via `--kubeconfig` to allow users to specify the _in-cluster context_ but pass credentials by setting a `ServiceAccount` for the `Pod` in question. 
+
+In order to opt into this, use the `"in-cluster"` string in a _context_ and omit a _user_:
```

Comment:
We don't want to allow using a different service account in cluster? This seems like a probable scenario, but we can just add it later